### ISSUE TITLE
HydroShare on Jupyter does not mark bag dirty

### ISSUE THREAD

**Issue by dtarb** [hydroshare/hydroshare#4765]
**Description of the bug**
I had a resource in HydroShare.  It was opened using CyberGIS Jupyter for water.  Then a new file created in Jupyter and uploaded to HydroShare was added to the resource.  This appears in the resource content area, but when I download the bag, it is missing, leading me to suspect that the bag was not marked dirty.  If I do something in the resource, such as change the abstract then download the bag - the new file is added.

**Comment by aaraney** [comment]
@dtarb, HS on Jupyter uses `hsclient` do all uploading and downloading. Im not familiar with the concept of a "dirty bag," but from your description, I am inferring that this is most likely an issue with `hsclient`, not hs-on-jupyter. Pinging @sblack-usu and @devincowan to triage.

**Comment by devincowan** [comment]
Looks like the [hsclient save()](https://github.com/hydroshare/hsclient/blob/master/hsclient/hydroshare.py#L236) uses an hsapi endpoint that [eventually should set the bag dirty](https://github.com/hydroshare/hydroshare/blob/develop/hs_file_types/utils.py#L293) but sounds like that's not happening correctly. @sblack-usu am I tracking that correctly?

**Comment by sblack-usu** [comment]
@devincowan - hsclient uses hsapi2 for updating metadata.  It looks like the metadata dirty flag [isn't being set](https://github.com/hydroshare/hydroshare/blob/develop/hs_rest_api2/metadata.py#L107).  Those two AVUs should be set. I think it should be set at the end of the `save_resource_metadata_xml` function.

**Comment by devincowan** [comment]
Thanks @sblack-usu on the hsapi2 correction. I will set dirty flag in that `save_resource_metadata_xml` function

**Comment by devincowan** [comment]
@sblack-usu re-looking at this, I don't see it using **hsapi2**
https://github.com/devincowan/hsclient/blob/master/hsclient/hydroshare.py#L350

```
class Resource(Aggregation):
     """Represents a Resource in HydroShare"""
     @property
     def _hsapi_path(self):
         path = urlparse(self.metadata.identifier).path
         return '/hsapi' + path
```

**Comment by devincowan** [comment]
@dtarb I haven't reproduced this yet
see this resource with a notebook that does a file upload:
https://www.hydroshare.org/resource/4253699e33314be99f78b4f477cdc417/
Can you share the resource (or the notebook) that had this issue?
I have pushed one small change that solves another related issue (and might solve your issue...I'm not sure until I can reproduce it).

**Comment by devincowan** [comment]
related to https://github.com/hydroshare/hsmodels/issues/37

**Comment by dtarb** [comment]
Steps I followed to reproduce
1. Create a resource on HydroShare.  (It is https://www.hydroshare.org/resource/2ef2450e290e470b9475b310f91fe5b4/.  I shared it with Developers group and invited Devin to this group.  Austin is already there.  So you have access)
2. Add a file (file1.txt)
3. Open resource with CUAHSI JupyterHub (using Python - v 3.8 server option)
4. In CUAHSI Jupyterhub add a second file (file2.txt)
5. Open HydroShare on Jupyter from the Launcher and Locate this resource (Date created sorting gets this to the top).  The interface shows that file 1.txt is synced but file2.txt is local to JupyterHub.
![image](https://user-images.githubusercontent.com/3456833/189506880-3950ca57-1469-4422-b2e9-abe8ae8b2398.png)
6. Upload file2.txt using HS on Jupyter (takes a while and progress is not indicated - so be patient)
7.  Verify back in HydroShare that both files are listed in the contents area of the resource
![image](https://user-images.githubusercontent.com/3456833/189506942-654e9cff-9a99-4455-ae2e-351267568522.png)

9. Download the bag and unzip it.  It only holds file1.txt.
![image](https://user-images.githubusercontent.com/3456833/189506954-6e85bcdb-9003-49c6-9122-9dd9822037d2.png)

If at this point I was to change the resource using HydroShare (e.g. edit the abstract), then if I download the bag, the bag is regenerated and I get both files.  But right now I am leaving this resource in the state where the bag only contains one file.

**Comment by sblack-usu** [comment]
@devincowan - you are right, it isn't using hsapi2.  It is using `/hsapi/resource/<resource_id>/ingest_metadata/`.  I'm looking at the code and the AVUs are set correctly.  https://github.com/hydroshare/hydroshare/blob/develop/hs_file_types/utils.py#L294

**Comment by sblack-usu** [comment]
@dtarb's description says that file upload is the issue and the bag is correct after metadata is updated.

**Comment by sblack-usu** [comment]
File upload should go through `/hsapi/resource/<resource_id>/files/` and that path looks like it sets the bag modified flag.  https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/resource_rest_api.py#L823

**Comment by devincowan** [comment]
yes @dtarb thanks for clarifying, I thought we were using hsclient to upload files from within a notebook, but I get that it is upload via hs-on-jupyter used in the workflow that creates the issue.

Thanks @aaraney !

**Comment by aaraney** [comment]
For reproducibility, below are the `hsclient`, `hsmodels`, and `hydroshare-on-jupyter` versions installed in the current (2022-09-13) Python - v3.8 CUAHSI Jupyter Hub server option (image)

```shell
hsclient==0.3.1
hsmodels==0.5.3
hydroshare-on-jupyter==0.1.5
```

**Comment by aaraney** [comment]
So, I was able to reproduce the issue in CUAHSI Jupyter Hub, however when I tried it locally using hydroshare-on-jupyter, I cannot reproduce the issue... digging into logs.

**Comment by aaraney** [comment]
Just reporting back, I didnt find anything substantive in the logs. @sblack-usu and I talked briefly near the end of the day about the issue and I caught him up on what `hydroshare-on-jupyter` is doing during its upload process. In brief, `hydroshare-on-jupyter` bundles to-be-uploaded files into a zip file, uploads that zip file to hydroshare, and then makes a call to hydroshare to unzip the file and overwrite any conflicting files. For clarity, this means if `file` is in the zip and `file` is _already_ in the resource, the _uploaded_ version of `file` will be persisted. This does not mean if an empty zip, for example, is uploaded `file` would not be deleted.

@sblack-usu is looking into hydroshare's unzip routine to see if this is where the bag is being dirtied.

**Comment by hyi** [comment]
@sblack-usu @devincowan @aaraney I am looking into this issue and it looks like this [unzip_file()](https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/utils.py#L1037) is called via REST API to unzip file while setting overwrite to be True (default is False). Is it right? If so, this routine has been updated substantially, but if overwrite is set to True, it should just overwrite files without raising exceptions. I assume default values for other parameters such as ```auto_aggregate``` and ```ingest_metadata``` are used without setting them differently. If so, it means it will auto aggregate if the unzipped files can be created into aggregations, and it will not ingest metadata. Note that if ingest_metadata is set to True, auto_aggregate has to be set to True and overwrite has to be set to True as well. Also setting ingest_metadata to True also delete the zip file by default, and setting ingest_metadata to True does trigger the ingest_metadata_files() which does set both metadata_dirty and bag_modified flags to True, so I bet if ingest_metadata is set to True, this issue will be resolved. So I am suggesting to try the following:
- Try with the latest release which includes my updated unzipping without creating an extra folder work to see if this issue would be resolved.
- If the issue were not resolved with the latest release, try to set ingest_metadata flag to be True when calling unzip_file() which should fix this issue.

**Comment by devincowan** [comment]
Thanks @hyi!
@aaraney and @sblack-usu how do we even test jupyterlab on local dev? I have JupyterLab running in a Docker container but don't know where to start to connect it to my local HS instance... Thanks!

**Comment by aaraney** [comment]
@devincowan its kind of a pain. I typically don't run JupyterLab locally when I am debugging, I generally run a local JupyterHub with whatever plugins I need installed in that environment. If you want a true debugging experience, you should be able to set breakpoints in your code of interest and setup your IDE to run `jupyter lab` when you run the debugger.

**Comment by hyi** [comment]
While thinking this through a bit more, I took my comment above back. I think as long as unzip file happens successfully, resource bag will be marked as dirty as shown in [this last line of the unzip() function](https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/utils.py#L1181). So not sure why downloaded bag does not contain new files.

**Comment by devincowan** [comment]
thanks @hyi 
@sblack-usu I was trying to get hydroshare-on-jupyter running in a local container because I wanted to see if these params make it through the post request:
https://github.com/hydroshare/hydroshare_on_jupyter/blob/master/hydroshare_on_jupyter/server.py#L577

**Comment by devincowan** [comment]
@sblack-usu here is a video that I took where I make a post request to `hsapi/resource/ID/functions/unzip/data/contents/FILENAME/` with the [same params used by hs_on_jupyter](https://github.com/hydroshare/hydroshare_on_jupyter/blob/master/hydroshare_on_jupyter/server.py#L577), then download the bag and it has the contents of the uzipped dir as it should...

https://user-images.githubusercontent.com/17934193/193854018-da1e2d1a-88d0-4258-aca5-33ef78cd27b1.mov

**Comment by devincowan** [comment]
@sblack-usu 

https://user-images.githubusercontent.com/17934193/193905177-a5a89eb5-8dea-4b31-bb56-cb9266ddf217.mov

same result: the file is in the bag

**Comment by sblack-usu** [comment]
I recreated this while looking at the access logs. I took notice of the `__zip.zip` filename, not sure if it matters yet.  The requests are 20 seconds apart, so the second request is happening after a successful upload.
```
34.133.87.169 - - [04/Oct/2022:19:22:56 +0000] "POST /hsapi/resource/cfc160172c5a4913a7e7e223913b75af/files/ HTTP/1.1" 201 98 "-" "python-requests/2.27.1"
34.133.87.169 - - [04/Oct/2022:19:23:16 +0000] "POST /hsapi/resource/cfc160172c5a4913a7e7e223913b75af/functions/unzip/data/contents/__zip.zip/ HTTP/1.1" 200 34 "-" "python-requests/2.27.1"
```

Relevant code block: https://github.com/hydroshare/hydroshare_on_jupyter/blob/9df912006b87d285f11a5a18d9c48ecf551b9c07/hydroshare_on_jupyter/server.py#L532-L547

I'd like to get my hands on the __zip.zip file.

**Comment by devincowan** [comment]
![image](https://user-images.githubusercontent.com/17934193/193926926-cc05539d-4829-40a8-875c-77edf6f3ac68.png)
Pylance isn't happy: `file_upload()` in hsclient [accepts the files as strings](https://github.com/devincowan/hsclient/blob/master/hsclient/hydroshare.py#L624)
But I don't think that's the issue

**Comment by sblack-usu** [comment]
That is a good point @devincowan, and I'm willing to bet It has something to do with the custom logic for zipping the files for upload.  .  [The lines above](https://github.com/hydroshare/hydroshare_on_jupyter/blob/9df912006b87d285f11a5a18d9c48ecf551b9c07/hydroshare_on_jupyter/server.py#L529) show it prepends a bagit prefix, which is `/data/contents/`.  I'm going to bet it has to do with that.  I want my hand on the zip file, but I can recreate the zip (with the bagit prefix) by hand.

**Comment by sblack-usu** [comment]
hmm, not so sure now...
```
                        # maintain file system structure relative to where data is stored in baggit (/data/contents/)
                        # Example: `/data/contents/dir1/some-file.txt` will be archived in the zip at, `/dir1/some-file.txt`
                        file.relative_to(resource_path_prefix / self.BAGGIT_PREFIX),
```

**Comment by sblack-usu** [comment]
I tried to start up the hs on jupyterhub project locally and had trouble running it.  I would like to get my hands on the zip file that is uploaded to hydroshare before it is unzipped.  I think it could give us a better idea of what is going on here.

This should be possible by comenting out the unzip command here.
https://github.com/hydroshare/hydroshare_on_jupyter/blob/9df912006b87d285f11a5a18d9c48ecf551b9c07/hydroshare_on_jupyter/server.py#L547

@devincowan or @Castronova - are either of you able to run the project locally and able to get that zip file?

**Comment by devincowan** [comment]
I haven't had any success running locally :(

**Comment by aaraney** [comment]
Hey guys! I hope you are doing well! Here is one of the [`__zip.zip`](https://github.com/hydroshare/hydroshare/files/9932483/__zip.zip) files for [this](https://www.hydroshare.org/resource/9a597bd5498e42679d038962b1ae7311/) hs resource. Hope this helps!

**Comment by aaraney** [comment]
Yeah that is super weird, @sblack-usu. Maybe try pulling the singleuser container cuahsi-jh is using and try to replicate it locally using that singleuser image? I don't think that will be overly difficult to do, don't quote me on that though haha. The singleuser image should be fine running as an isolated jupyter lab instance running outside of a jupyter hub environment. I can try to help with that if you can get me access to pull the singleuser image.

**Comment by sblack-usu** [comment]
The only other thing that feels like a possibility is that hs on jh makes a follow-up request that resets the bag dirty flag to false.  I'm very doubtful of that, but it is the most plausible scenario I can currently think of.

**Comment by sblack-usu** [comment]
I looked through hydroshare code, and the only time the flag is set back to false is when the bag is generated.

**Comment by dtarb** [comment]
Chiming in from perhaps a naive perspective, but what happens in the HydroShare code between the zip upload and unzip actions and the code to actually mark the bag dirty.  If a failure is occurring there (in hydroshare), it may result in the bag not being marked dirty, but the action looking like it was done right.  Is there something different (in terms of permissions perhaps) with this code running triggered from CUAHSI JupyterHub than in your tests.  Feel free to ignore or dismiss if this comment reflects my lack of understanding as to how this really works.

**Comment by aaraney** [comment]
> Is there something different (in terms of permissions perhaps) with this code running triggered from CUAHSI JupyterHub than in your tests.

You might be onto something, @dtarb. I wonder if there is something weird going on with temporary directories on CUAHSI Jupyterhub. @sblack-usu, have you tried running your above code on CUAHSI Jupyterhub?

**Comment by sblack-usu** [comment]
I have, but that isn't necessarily the same environment that hs on js runs on, is it?  I guess I'm not entirely sure how plugins work on jupyterhub, I don't think it runs on the single user image, does it?

I'm not seeing any errors and looking at the code if the zip is successfully unzipped, an error after that would be raised be raised with a 400+ status code in the response.  I tested with the same user and even the same type of authorization.

**Comment by aaraney** [comment]
Yeah, it is the same environment. JupyterHub creates instances of the singleuser image which is just a JupyterLab image with some extra communication bells and whistles. The singleuser image is where hs on jupyter is installed and configured.

**Comment by devincowan** [comment]
Discussed today. I'm going to take another look at this with some fresh eyes...

**Comment by devincowan** [comment]
related to https://github.com/hydroshare/hsclient/issues/56
hsclient somehow toggles the dirty_bag_flag without regenerating the zip
